### ISSUE TITLE
[强化学习组合再平衡] | RL Portfolio Rebalancing

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4471]
**标签**: quantitative-strategy
### 摘要
使用深度强化学习（PPO算法）动态学习市场状态，实现组合再平衡的最优时机选择与仓位优化，突破传统固定阈值再平衡的局限性。

### 算法描述

**核心原理：**

传统再平衡策略依赖固定阈值或固定时间窗口，无法适应市场状态的动态变化。强化学习（RL）通过与市场环境交互，学习在不同市场状态下选择最优再平衡时机和目标权重。

**数学模型：**

1. **马尔可夫决策过程（MDP）**：
- **状态空间**：$S_t = (w_t, r_t, \sigma_t, VIX_t, trend_t)$
  - $w_t$：当前权重向量
  - $r_t$：近期收益率
  - $\sigma_t$：波动率
  - $VIX_t$：恐慌指数
  - $trend_t$：趋势指标

- **动作空间**：$A_t = \{\Delta w_1, \Delta w_2, ..., \Delta w_n\}$
  - 各资产目标权重调整量

- **奖励函数**：
$$R_t = \alpha \cdot \Delta w_t^T r_t - \beta \cdot TC_t - \gamma \cdot \text{risk_penalty}$$

其中 $TC_t$ 为交易成本，$\alpha, \beta, \gamma$ 为权重系数。

2. **PPO算法核心**：

- **Clip机制**：
$$L^{CLIP}(\theta) = \mathbb{E}\left[\min(r_t(\theta) \cdot A_t, \text{clip}(r_t(\theta), 1-\epsilon, 1+\epsilon) \cdot A_t)\right]$$

- **价值函数估计**：
$$V_{\theta_v}(s_t) \approx \mathbb{E}\left[\sum_{i=t}^{T} \gamma^{i-t} R_i\right]$$

- **熵正则化**：
$$H(\pi_\theta) = -\sum_a \pi_\theta(a|s) \log \pi_\theta(a|s)$$

3. **市场状态特征**：
$$M_t = \text{sign}(r_t) \cdot |r_t|^{\frac{1}{4}} \cdot (1 + \sigma_t)^{-1}$$

4. **最优再平衡时机**：
当策略网络输出的动作 $|A_t| > \epsilon_{action}$ 时执行再平衡。

**交易流程：**

1. **环境初始化**：加载历史市场数据，定义状态空间和动作空间
2. **PPO训练**：迭代训练策略网络和价值网络
3. **在线Inference**：实时获取市场状态，输入策略网络，输出再平衡动作
4. **执行与学习**：执行动作后获得奖励，更新网络参数

### Python代码示例

```python
"""
强化学习组合再平衡策略
RL Portfolio Rebalancing Strategy

核心逻辑：
1. 构建MDP市场环境
2. 使用PPO算法训练策略网络
3. 在线推理执行最优再平衡

作者：Ginkgo量化研究组
版本：v1.0
"""

import numpy as np
import pandas as pd
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from enum import Enum
import gym
from gym import spaces
import torch
import torch.nn as nn
import torch.optim as optim
from torch.distributions import Normal


class MarketRegime(Enum):
    """市场状态枚举"""
    BULL = "bull"
    BEAR = "bear"
    VOLATILE = "volatile"
    STABLE = "stable"


@dataclass
class PortfolioState:
    """组合状态"""
    weights: np.ndarray          # 当前权重
    returns: np.ndarray          # 近期收益率
    volatility: float            # 波动率
    vix: float                   # VIX指数
    trend: float                 # 趋势指标
    regime: MarketRegime         # 市场状态


class PortfolioRebalanceEnv(gym.Env):
    """组合再平衡强化学习环境"""
    
    def __init__(
        self,
        price_data: pd.DataFrame,
        initial_capital: float = 1000000,
        transaction_cost: float = 0.001,
        risk_penalty: float = 0.1
    ):
        """
        初始化环境
        
        Args:
            price_data: 价格历史数据
            initial_capital: 初始资金
            transaction_cost: 交易成本比例
            risk_penalty: 风险惩罚系数
        """
        super().__init__()
        
        self.price_data = price_data
        self.initial_capital = initial_capital
        self.transaction_cost = transaction_cost
        self.risk_penalty = risk_penalty
        
        self.n_assets = len(price_data.columns)
        self.n_steps = len(price_data)
        
        # 动作空间：权重调整量（连续）
        self.action_space = spaces.Box(
            low=-1.0, high=1.0, shape=(self.n_assets,), dtype=np.float32
        )
        
        # 状态空间：权重 + 收益率 + 波动率 + VIX + 趋势
        self.observation_space = spaces.Box(
            low=-np.inf, high=np.inf, 
            shape=(self.n_assets * 3 + 2,), dtype=np.float32
        )
        
        self.current_step = 0
        self.current_weights = np.ones(self.n_assets) / self.n_assets
        self.current_capital = initial_capital
        
    def reset(self) -> np.ndarray:
        """重置环境"""
        self.current_step = 0
        self.current_weights = np.ones(self.n_assets) / self.n_assets
        self.current_capital = self.initial_capital
        return self._get_observation()
    
    def _get_observation(self) -> np.ndarray:
        """获取当前状态"""
        if self.current_step == 0:
            returns = np.zeros(self.n_assets)
        else:
            returns = self.price_data.iloc[self.current_step].values / \
                     self.price_data.iloc[max(0, self.current_step-1)].values - 1
        
        volatility = self._calculate_volatility()
        vix = self._get_vix()
        trend = self._get_trend()
        
        obs = np.concatenate([
            self.current_weights,
            returns,
            [volatility, vix, trend]
        ])
        return obs.astype(np.float32)
    
    def _calculate_volatility(self, window: int = 20) -> float:
        """计算历史波动率"""
        if self.current_step < window:
            return 0.02
        
        returns = []
        for i in range(self.current_step - window, self.current_step):
            if i > 0:
                ret = self.price_data.iloc[i].values / \
                      self.price_data.iloc[i-1].values - 1
                returns.append(np.mean(ret))
        
        return np.std(returns) if len(returns) > 1 else 0.02
    
    def _get_vix(self) -> float:
        """模拟VIX值"""
        return 20.0 + np.random.normal(0, 5)
    
    def _get_trend(self) -> float:
        """计算趋势指标（简单移动平均差）"""
        if self.current_step < 20:
            return 0.0
        
        short_ma = []
        long_ma = []
        
        for i in range(self.n_assets):
            prices = self.price_data.iloc[:, i]
            short_ma.append(prices.iloc[max(0, self.current_step-5):self.current_step].mean())
            long_ma.append(prices.iloc[max(0, self.current_step-20):self.current_step].mean())
        
        short_ma = np.array(short_ma)
        long_ma = np.array(long_ma)
        
        return np.mean((short_ma - long_ma) / long_ma)
    
    def step(self, action: np.ndarray) -> Tuple[np.ndarray, float, bool, Dict]:
        """
        执行动作
        
        Args:
            action: 权重调整量
        
        Returns:
            (next_state, reward, done, info)
        """
        # 解析动作为目标权重
        target_weights = self.current_weights + action * 0.1
        target_weights = np.clip(target_weights, 0.01, 0.99)
        target_weights = target_weights / target_weights.sum()
        
        # 计算交易成本
        weight_diff = np.abs(target_weights - self.current_weights)
        turnover = weight_diff.sum() / 2
        tc = turnover * self.transaction_cost * self.current_capital
        
        # 执行交易
        self.current_weights = target_weights
        
        # 计算下一步收益
        self.current_step += 1
        
        if self.current_step >= self.n_steps:
            done = True
            reward = 0
            return self._get_observation(), reward, done, {}
        
        # 计算组合收益
        current_prices = self.price_data.iloc[self.current_step].values
        prev_prices = self.price_data.iloc[max(0, self.current_step-1)].values
        asset_returns = current_prices / prev_prices - 1
        
        portfolio_return = np.dot(self.current_weights, asset_returns)
        
        # 风险惩罚（波动率）
        volatility_penalty = self._calculate_volatility() * self.risk_penalty
        
        # 奖励 = 组合收益 - 交易成本 - 风险惩罚
        reward = portfolio_return * self.current_capital - tc - \
                volatility_penalty * self.current_capital
        
        self.current_capital *= (1 + portfolio_return)
        
        done = self.current_step >= self.n_steps - 1
        
        info = {
            "portfolio_value": self.current_capital,
            "turnover": turnover,
            "portfolio_return": portfolio_return
        }
        
        return self._get_observation(), reward, done, info


class PPOActor(nn.Module):
    """PPO策略网络"""
    
    def __init__(self, state_dim: int, action_dim: int, hidden_dim: int = 128):
        super().__init__()
        
        self.net = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, action_dim),
            nn.Tanh()  # 输出[-1, 1]
        )
        
        self.log_std = nn.Parameter(torch.zeros(action_dim))
    
    def forward(self, state: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
        """前向传播"""
        mean = self.net(state)
        log_std = self.log_std.expand_as(mean)
        std = torch.exp(log_std)
        return mean, std


class PPOCritic(nn.Module):
    """PPO价值网络"""
    
    def __init__(self, state_dim: int, hidden_dim: int = 128):
        super().__init__()
        
        self.net = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, 1)
        )
    
    def forward(self, state: torch.Tensor) -> torch.Tensor:
        return self.net(state)


class PPOAgent:
    """PPO强化学习智能体"""
    
    def __init__(
        self,
        state_dim: int,
        action_dim: int,
        lr_actor: float = 3e-4,
        lr_critic: float = 1e-3,
        gamma: float = 0.99,
        epsilon: float = 0.2,
        epochs: int = 10
    ):
        """
        初始化PPO智能体
        
        Args:
            state_dim: 状态维度
            action_dim: 动作维度
            lr_actor: 策略网络学习率
            lr_critic: 价值网络学习率
            gamma: 折扣因子
            epsilon: Clip参数
            epochs: 训练轮数
        """
        self.gamma = gamma
        self.epsilon = epsilon
        self.epochs = epochs
        
        self.actor = PPOActor(state_dim, action_dim)
        self.critic = PPOCritic(state_dim)
        
        self.actor_optimizer = optim.Adam(self.actor.parameters(), lr=lr_actor)
        self.critic_optimizer = optim.Adam(self.critic.parameters(), lr=lr_critic)
    
    def select_action(self, state: np.ndarray) -> np.ndarray:
        """选择动作"""
        state_tensor = torch.FloatTensor(state).unsqueeze(0)
        
        with torch.no_grad():
            mean, std = self.actor(state_tensor)
            dist = Normal(mean, std)
            action = dist.sample()
            action = torch.clamp(action, -1, 1)
        
        return action.numpy()[0]
    
    def update(
        self,
        states: torch.Tensor,
        actions: torch.Tensor,
        old_log_probs: torch.Tensor,
        rewards: torch.Tensor,
        dones: torch.Tensor
    ):
        """
        更新策略网络
        
        Args:
            states: 状态序列
            actions: 动作序列
            old_log_probs: 旧的对数概率
            rewards: 奖励序列
            dones: 终止标记
        """
        # 计算GAE
        values = self.critic(states).squeeze()
        advantages = []
        gae = 0
        
        for t in reversed(range(len(rewards))):
            if t == len(rewards) - 1:
                next_value = 0
            else:
                next_value = values[t + 1].item()
            
            delta = rewards[t] + self.gamma * next_value * (1 - dones[t]) - values[t].item()
            gae = delta + self.gamma * gae * (1 - dones[t])
            advantages.insert(0, gae)
        
        advantages = torch.FloatTensor(advantages)
        returns = advantages + values.detach()
        
        # 归一化优势函数
        advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
        
        for _ in range(self.epochs):
            # 计算新的动作概率
            mean, std = self.actor(states)
            dist = Normal(mean, std)
            new_log_probs = dist.log_prob(actions).sum(dim=-1)
            
            # 计算比率和剪辑损失
            ratio = torch.exp(new_log_probs - old_log_probs)
            
            surr1 = ratio * advantages
            surr2 = torch.clamp(ratio, 1 - self.epsilon, 1 + self.epsilon) * advantages
            
            actor_loss = -torch.min(surr1, surr2).mean()
            
            # 价值损失
            critic_loss = nn.functional.mse_loss(values, returns)
            
            # 更新网络
            self.actor_optimizer.zero_grad()
            actor_loss.backward()
            self.actor_optimizer.step()
            
            self.critic_optimizer.zero_grad()
            critic_loss.backward()
            self.critic_optimizer.step()


class RLPortfolioRebalanceStrategy:
    """强化学习组合再平衡策略"""
    
    def __init__(
        self,
        price_data: pd.DataFrame,
        initial_capital: float = 1000000,
        transaction_cost: float = 0.001
    ):
        """
        初始化策略
        
        Args:
            price_data: 价格数据
            initial_capital: 初始资金
            transaction_cost: 交易成本
        """
        self.price_data = price_data
        self.initial_capital = initial_capital
        self.transaction_cost = transaction_cost
        
        # 创建环境
        self.env = PortfolioRebalanceEnv(
            price_data, initial_capital, transaction_cost
        )
        
        # 创建智能体
        state_dim = self.env.observation_space.shape[0]
        action_dim = self.env.action_space.shape[0]
        
        self.agent = PPOAgent(
            state_dim=state_dim,
            action_dim=action_dim,
            lr_actor=3e-4,
            lr_critic=1e-3,
            gamma=0.99,
            epsilon=0.2
        )
        
        self.weights_history = []
    
    def train(self, n_episodes: int = 100, batch_size: int = 32):
        """
        训练PPO智能体
        
        Args:
            n_episodes: 训练回合数
            batch_size: 批大小
        """
        print("开始训练PPO智能体...")
        
        for episode in range(n_episodes):
            state = self.env.reset()
            episode_reward = 0
            states, actions, rewards, dones, log_probs = [], [], [], [], []
            
            while True:
                # 选择动作
                action = self.agent.select_action(state)
                
                # 获取动作的对数概率
                state_tensor = torch.FloatTensor(state).unsqueeze(0)
                with torch.no_grad():
                    mean, std = self.agent.actor(state_tensor)
                    dist = Normal(mean, std)
                    log_prob = dist.log_prob(torch.FloatTensor(action)).sum(dim=-1)
                
                # 执行动作
                next_state, reward, done, info = self.env.step(action)
                
                # 记录经验
                states.append(state)
                actions.append(action)
                rewards.append(reward)
                dones.append(done)
                log_probs.append(log_prob.item())
                
                episode_reward += reward
                state = next_state
                
                if done:
                    break
            
            # 批量更新
            if episode % batch_size == 0 and episode > 0:
                states_t = torch.FloatTensor(np.array(states))
                actions_t = torch.FloatTensor(np.array(actions))
                rewards_t = torch.FloatTensor(np.array(rewards))
                dones_t = torch.FloatTensor(np.array(dones))
                old_log_probs_t = torch.FloatTensor(np.array(log_probs))
                
                self.agent.update(
                    states_t, actions_t, old_log_probs_t,
                    rewards_t, dones_t
                )
            
            if episode % 10 == 0:
                print(f"Episode {episode}: Reward = {episode_reward:.2f}, "
                      f"Capital = {info['portfolio_value']:.2f}")
    
    def run_inference(self) -> pd.DataFrame:
        """运行推理"""
        state = self.env.reset()
        results = []
        
        while True:
            action = self.agent.select_action(state)
            next_state, reward, done, info = self.env.step(action)
            
            self.weights_history.append(self.env.current_weights.copy())
            
            results.append({
                "portfolio_value": info["portfolio_value"],
                "weights": self.env.current_weights.copy(),
                "portfolio_return": info["portfolio_return"]
            })
            
            state = next_state
            
            if done:
                break
        
        return pd.DataFrame(results)


def generate_synthetic_portfolio_data(
    n_assets: int = 5,
    n_days: int = 252
) -> pd.DataFrame:
    """生成合成组合数据"""
    np.random.seed(42)
    
    assets = ['Asset_' + chr(65 + i) for i in range(n_assets)]
    data = {}
    
    for asset in assets:
        initial_price = np.random.uniform(80, 120)
        returns = np.random.normal(0.0005, 0.015, n_days)
        prices = initial_price * np.exp(np.cumsum(returns))
        data[asset] = prices
    
    return pd.DataFrame(data)


if __name__ == "__main__":
    print("=" * 60)
    print("强化学习组合再平衡策略")
    print("=" * 60)
    
    # 生成数据
    price_data = generate_synthetic_portfolio_data(n_assets=5, n_days=252)
    
    # 创建策略
    strategy = RLPortfolioRebalanceStrategy(
        price_data=price_data,
        initial_capital=1000000,
        transaction_cost=0.001
    )
    
    # 训练
    strategy.train(n_episodes=100)
    
    # 推理
    results = strategy.run_inference()
    
    print("\n【推理结果】")
    print(f"初始资金: {1000000:.2f}")
    print(f"最终资金: {results['portfolio_value'].iloc[-1]:.2f}")
    print(f"总收益率: {(results['portfolio_value'].iloc[-1] / 1000000 - 1) * 100:.2f}%")
